packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js

```diff
@@ -14,43 +19,196 @@ const {
+    time_to_archive: getGranuleTimeToArchive(granule),
+    collection_cumulus_id: collectionCumulusId,
+    provider_cumulus_id: providerCumulusId,
+    execution_cumulus_id: executionCumulusId,
+    pdr_cumulus_id: pdrCumulusId,
+    // Temporal info from CMR
+    beginning_date_time: temporalInfo.beginningDateTime,
+    ending_date_time: temporalInfo.endingDateTime,
+    production_date_time: temporalInfo.productionDateTime,
+    last_update_date_time: temporalInfo.lastUpdateDateTime,
+    // Processing info from execution
+    processing_start_date_time: processingTimeInfo.processingStartDateTime,
+    processing_end_date_time: processingTimeInfo.processingEndDateTime,
+  };
+};
+
+const generateFileRecord = ({ file, granuleCumulusId }) => ({
+  ...translateApiFiletoPostgresFile(file),
+  granule_cumulus_id: granuleCumulusId,
+});
+
+const generateFileRecords = async ({
+  files,
+  granuleCumulusId,
+}) => files.map((file) => generateFileRecord({ file, granuleCumulusId }));
+
+const writeFilesViaTransaction = async ({
+  fileRecords,
   trx,
+  filePgModel = new FilePgModel(),
 }) =>
-  trx(tableNames.granules)
-    .insert({
-      granule_id: granule.granuleId,
-      status: getGranuleStatus(cumulusMessage, granule),
-      collection_cumulus_id: collectionCumulusId,
-      provider_cumulus_id: providerCumulusId,
-      execution_cumulus_id: executionCumulusId,
-      pdr_cumulus_id: pdrCumulusId,
-    });
+  Promise.all(fileRecords.map(
+    (fileRecord) => filePgModel.upsert(trx, fileRecord)
+  ));
+
+/**
+ * Get the cumulus ID from a query result or look it up in the database.
+ *
+ * For certain cases, such as an upsert query that matched no rows, an empty
+ * database result is returned, so no cumulus ID will be returned. In those
+ * cases, this function will lookup the granule cumulus ID from the record.
+ *
+ * @param {Object} params
+ * @param {Object} params.trx - A Knex transaction
+ * @param {Object} params.queryResult - Query result
+ * @param {Object} params.granuleRecord - A granule record
+ * @param {Object} params.granulePgModel - Database model for granule data
+ * @returns {Promise<number|undefined>} - Cumulus ID for the granule record
+ */
+const getGranuleCumulusIdFromQueryResultOrLookup = async ({
+  queryResult = [],
+  granuleRecord,
+  trx,
+  granulePgModel = new GranulePgModel(),
+}) => {
+  let [granuleCumulusId] = queryResult;
+  if (!granuleCumulusId) {
+    granuleCumulusId = await granulePgModel.getRecordCumulusId(
+      trx,
+      { granule_id: granuleRecord.granule_id }
+    );
+  }
+  return granuleCumulusId;
+};
+
+const writeGranuleAndFilesViaTransaction = async ({
+  cumulusMessage,
+  granule,
+  processingTimeInfo,
+  provider,
+  collectionCumulusId,
+  providerCumulusId,
+  executionCumulusId,
+  pdrCumulusId,
+  granulePgModel = new GranulePgModel(),
+  fileUtils = FileUtils,
+  trx,
+}) => {
+  const { files = [] } = granule;
+  // TODO: I think this is necessary to set properties like
```

Comment:
I left this is in because I wanted someone to sanity check me. From what I can tell, we have no guarantee that the file objects in `granule.files` for an incoming Cumulus message have the properties `size`, `key`, and possibly others. But for `size`, it is at least required to calculate the granule object `productVolume`. And `key` isn't required on the PG file schema yet, though I think it perhaps should be

Regardless, does this approach seem like the right one? I think that as long as the contents of `granule.files` are unpredictable, we have to do something like this, but I could be missing something

Reply:
Where would that guarantee come from *other* than here?

Reply:
1ab799021aaa8ca689cd0276f75e54f51df7c737